File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n=nums.length;
        int res[]=new int[n];
        Stack<Integer> ss=new Stack<>();
        for(int i=2*n-1;i>=0;i--)
        {
            while(!ss.isEmpty()&&nums[ss.peek()]<=nums[(i%n)])
            {
                ss.pop();
            }
            if(ss.isEmpty())
            {
                res[i%n]=-1;
            }
            else{
                res[(i%n)]=nums[ss.peek()];
            }
            ss.push(i%n);
        }
        return res;
    }
}

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n=len(nums)
        res=[-1]*n
        st=[]
        for i in range(2*n-1, -1, -1):
            v=nums[i%n]
            while st and st[-1]<=v:
                st.pop()
            if st:
                res[i%n]=st[-1]
            st.append(v)
        return res

        